From 20be15184a5a3854a6151b8e1c13257ae4cd7d04 Mon Sep 17 00:00:00 2001
From: "Hebbar, Gururaja" <gururaja.hebbar@ti.com>
Date: Tue, 28 Feb 2012 14:10:27 +0530
Subject: [PATCH 132/193] ARM: OMAP2+: mcasp: use omap_device api for
 registration

Convert the old-style device registration code for mcasp to use
omap_device.  This will allow the driver to be converted to use PM
runtime and to take advantage of the OMAP IP block management
infrastructure (hwmod, PM, etc.).

Signed-off-by: Hebbar, Gururaja <gururaja.hebbar@ti.com>
---
 arch/arm/mach-omap2/devices.c |   54 ++++++++++++++++++-----------------------
 arch/arm/mach-omap2/devices.h |    2 +-
 2 files changed, 24 insertions(+), 32 deletions(-)

diff --git a/arch/arm/mach-omap2/devices.c b/arch/arm/mach-omap2/devices.c
index 5f4e3ef..c89e11d 100644
--- a/arch/arm/mach-omap2/devices.c
+++ b/arch/arm/mach-omap2/devices.c
@@ -196,42 +196,34 @@ int __init am33xx_register_tsc(struct tsc_data *pdata)
 
 #if defined(CONFIG_SND_AM335X_SOC_EVM) || \
 				defined(CONFIG_SND_AM335X_SOC_EVM_MODULE)
-static struct resource am335x_mcasp1_resource[] = {
-	{
-		.name = "mcasp1",
-		.start = AM33XX_ASP1_BASE,
-		.end = AM33XX_ASP1_BASE + (SZ_1K * 12) - 1,
-		.flags = IORESOURCE_MEM,
-	},
-	/* TX event */
-	{
-		.start = AM33XX_DMA_MCASP1_X,
-		.end = AM33XX_DMA_MCASP1_X,
-		.flags = IORESOURCE_DMA,
-	},
-	/* RX event */
-	{
-		.start = AM33XX_DMA_MCASP1_R,
-		.end = AM33XX_DMA_MCASP1_R,
-		.flags = IORESOURCE_DMA,
-	},
-};
+int __init am335x_register_mcasp(struct snd_platform_data *pdata, int ctrl_nr)
+{
+	int l;
+	struct omap_hwmod *oh;
+	struct platform_device *pdev;
+	char oh_name[12];
+	char *dev_name = "davinci-mcasp";
 
-static struct platform_device am335x_mcasp1_device = {
-	.name = "davinci-mcasp",
-	.id = 1,
-	.num_resources = ARRAY_SIZE(am335x_mcasp1_resource),
-	.resource = am335x_mcasp1_resource,
-};
+	l = snprintf(oh_name, 12, "mcasp%d", ctrl_nr);
 
-void __init am335x_register_mcasp1(struct snd_platform_data *pdata)
-{
-	am335x_mcasp1_device.dev.platform_data = pdata;
-	platform_device_register(&am335x_mcasp1_device);
+	oh = omap_hwmod_lookup(oh_name);
+	if (!oh) {
+		pr_err("could not look up %s\n", oh_name);
+		return -ENODEV;
+	}
+
+	pdev = omap_device_build(dev_name, ctrl_nr, oh, pdata,
+			sizeof(struct snd_platform_data), NULL, 0, 0);
+	WARN(IS_ERR(pdev), "Can't build omap_device for %s:%s.\n",
+			dev_name, oh->name);
+	return IS_ERR(pdev) ? PTR_ERR(pdev) : 0;
 }
 
 #else
-void __init am335x_register_mcasp1(struct snd_platform_data *pdata) {}
+int __init am335x_register_mcasp(struct snd_platform_data *pdata, int ctrl_nr)
+{
+	return 0;
+}
 #endif
 
 #if (defined(CONFIG_SND_AM33XX_SOC) || (defined(CONFIG_SND_AM33XX_SOC_MODULE)))
diff --git a/arch/arm/mach-omap2/devices.h b/arch/arm/mach-omap2/devices.h
index f421d6f..8174cf3 100644
--- a/arch/arm/mach-omap2/devices.h
+++ b/arch/arm/mach-omap2/devices.h
@@ -16,7 +16,7 @@ struct isp_platform_data;
 
 int omap3_init_camera(struct isp_platform_data *pdata);
 
-void __init am335x_register_mcasp1(struct snd_platform_data *pdata);
+int __init am335x_register_mcasp(struct snd_platform_data *pdata, int ctrl_nr);
 extern int __init am33xx_register_tsc(struct tsc_data *pdata);
 extern void register_ehrpwm(int max_freq);
 
-- 
1.7.9.5

